zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/znode/groupacl/X509ZNodeGroupAclProvider.java

```diff
@@ -0,0 +1,180 @@
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.common.ZKConfig;
+import org.apache.zookeeper.data.Id;
+import org.apache.zookeeper.server.ServerCnxn;
+import org.apache.zookeeper.server.ZooKeeperServer;
+import org.apache.zookeeper.server.auth.ServerAuthenticationProvider;
+import org.apache.zookeeper.server.auth.X509AuthenticationUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A ServerAuthenticationProvider implementation that does both authentication and authorization for protecting znodes from unauthorized access.
+ * Znodes are grouped into domains according to their ownership, and clients are granted access permission to domains.
+ * Authentication mechanism is same as in X509AuthenticationProvider. If authentication is failed,
+ *  decline the connection request.
+ * Authorization is done by checking clientId which is usually an URI (uniform resource identifier)in ACL metadata for matched domains.
+ * Detailed step for authorization is:
+ *  1. Acl provider attempts to extract the clientId from the cert provided by the client
+ *  2. Acl provider attempts to look up the client's domain using the clientId and ClientUriDomainMappingHelper
+ *  3. If matched domain is found, add the domain in authInfo in the connection object;
+ *     if no matched domain is found, add the clientId in authInfo instead; either way, establish the connection.
+ * This class is meant to support the following use patterns:
+ *  1. Single domain: The creator and the accessor to the znodes belong to the same domain. This is
+ *     the most straightforward use case, and accessors from other domains will be declined to access the nodes.
+ *  2. Super user domain: The accessors need permission to access znodes created by creators belong
+ *     to many other domains. In this case the accessors will be mapped to super user domain and be
+ *     given super user privilege.
+ *  3. Open read access: The znodes need to be accessed by accessors from many different domains, so when
+ *     the creators create these znodes, these nodes will be given "open read access" (see below).
+ * Optional features include:
+ *    "auto-set ACL": add ZooDefs.Ids.CREATOR_ALL_ACL to all newly-created znodes,
+ *    "open read access": add (world, anyone r) to all newly-written znodes whose path prefixes are
+ *          given in the znode group acl config (comma-delimited, multiple such prefixes are possible).
+ */
+public class X509ZNodeGroupAclProvider extends ServerAuthenticationProvider {
+  private static final Logger LOG = LoggerFactory.getLogger(X509ZNodeGroupAclProvider.class);
+  private final String logStrPrefix = this.getClass().getName() + ":: ";
+  private final X509TrustManager trustManager;
+  private final X509KeyManager keyManager;
+  static final String ZOOKEEPER_ZNODEGROUPACL_SUPERUSER = "zookeeper.znodeGroupAcl.superUser";
+  private static ClientUriDomainMappingHelper uriDomainMappingHelper = null;
```

Comment:
why is this static again? this should ***not*** be static. 

Please do this:

`private volatile ClientUriDomainMappingHelper uriDomainMappingHelper = null;`

Reply:
Regarding `volatile` - using `volatile` can prevent creating multiple objects which is unlikely, but I have a concern on the overhead of `volatile` on the performance of read the object. Discussed offline with @narendly, decided not to use `volatile`.